group-coordinator/src/main/java/org/apache/kafka/coordinator/group/share/ShareCoordinatorService.java

```diff
@@ -18,18 +18,152 @@
+      this.writer = writer;
+      return this;
+    }
+
+    public Builder withLoader(CoordinatorLoader<Record> loader) {
+      this.loader = loader;
+      return this;
+    }
+
+    public Builder withTime(Time time) {
+      this.time = time;
+      return this;
+    }
+
+    public Builder withTimer(Timer timer) {
+      this.timer = timer;
+      return this;
+    }
+
+    public Builder withCoordinatorMetrics(ShareCoordinatorMetrics coordinatorMetrics) {
+      this.coordinatorMetrics = coordinatorMetrics;
+      return this;
+    }
+
+    public Builder withCoordinatorRuntimeMetrics(CoordinatorRuntimeMetrics coordinatorRuntimeMetrics) {
+      this.coordinatorRuntimeMetrics = coordinatorRuntimeMetrics;
+      return this;
+    }
+
+    public ShareCoordinatorService build() {
+      //todo maybe move to common class as similar to GroupCoordinatorService
+      if (config == null) {
+        throw new IllegalArgumentException("Config must be set.");
+      }
+      if (writer == null) {
+        throw new IllegalArgumentException("Writer must be set.");
+      }
+      if (loader == null) {
+        throw new IllegalArgumentException("Loader must be set.");
+      }
+      if (time == null) {
+        throw new IllegalArgumentException("Time must be set.");
+      }
+      if (timer == null) {
+        throw new IllegalArgumentException("Timer must be set.");
+      }
+      if (coordinatorMetrics == null) {
+        throw new IllegalArgumentException("Share Coordinator metrics must be set.");
+      }
+      if (coordinatorRuntimeMetrics == null) {
+        throw new IllegalArgumentException("Coordinator runtime metrics must be set.");
+      }
+
+      String logPrefix = String.format("ShareCoordinator id=%d", nodeId);
+      LogContext logContext = new LogContext(String.format("[%s] ", logPrefix));
+
+      CoordinatorShardBuilderSupplier<ShareCoordinatorShard, Record> supplier = () ->
+          new ShareCoordinatorShard.Builder(config);
+
+      CoordinatorEventProcessor processor = new MultiThreadedEventProcessor(
+          logContext,
+          "share-coordinator-event-processor-",
+          config.numThreads,
+          time,
+          coordinatorRuntimeMetrics
+      );
+
+      CoordinatorRuntime<ShareCoordinatorShard, Record> runtime =
+          new CoordinatorRuntime.Builder<ShareCoordinatorShard, Record>()
+              .withTime(time)
+              .withTimer(timer)
+              .withLogPrefix(logPrefix)
+              .withLogContext(logContext)
+              .withEventProcessor(processor)
+              .withPartitionWriter(writer)
+              .withLoader(loader)
+              .withCoordinatorShardBuilderSupplier(supplier)
+              .withTime(time)
+              .withDefaultWriteTimeOut(Duration.ofMillis(config.writeTimeoutMs))
+              .withCoordinatorRuntimeMetrics(coordinatorRuntimeMetrics)
+              .withCoordinatorMetrics(coordinatorMetrics)
+              .build();
```

Comment:
Do we require another coordinator runtime or could share the runtime with group coordinator. There might be some additional work to share it but enevtually that's the event processor. Wdyt?

Reply:
We cannot share since the runtime instantiation requires concrete type parameter of `CoordinatorShard` type. The implementations for group coord and share coord are different.

Reply:
I understand that but was just thinking that will be it better to have a single runtime. Though read the class spec for CoordinatorRuntime again and it does say to have different coordinator implementations. Yeah let's have different one itself.

Reply:
@apoorvmittal10 we can revisit if there are significant perf improvements to be had